#164797150 users can rate article

nwamugo opened the pull request:
#### What does this PR do?
This PR allows authenticated users to rate an article by giving it a number.

#### Description of Task to be completed?
User is able to rate articles authored by others

#### How should this be manually tested?
Clone the repo
npm install the packages
Open postman, login or signup
Include x-access-token, and a json object in the body { rating: 3 }
post to the route localhost:3000/api/v1/articles/141f4f05-7d81-4593-ab54-e256c1006210/rate

#### Any background context you want to provide?
The community of authors is built on trust. Rating each others articles helps to gain and build trust. Leaving a rating is an author's chance to give other authors a general idea of what they feel about their content

#### What are the relevant pivotal tracker stories?(if applicable)
#164797150 Users should be able to give a rating to an article

#### Screenshots
![image](https://user-images.githubusercontent.com/41264503/56248908-c1dd0780-60a1-11e9-97e3-94fcca4ca758.png)

#### Flowchart
![image](https://user-images.githubusercontent.com/41264503/56249160-aa524e80-60a2-11e9-9500-a596d994db67.png)

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
parent 1dac2e654512dde9a2cf2ad4a7a5cc36e5d44df9
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556211095 +0100

parent 1dac2e654512dde9a2cf2ad4a7a5cc36e5d44df9
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556210817 +0100

parent 1dac2e654512dde9a2cf2ad4a7a5cc36e5d44df9
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556210679 +0100

parent 1dac2e654512dde9a2cf2ad4a7a5cc36e5d44df9
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556210651 +0100

parent 1dac2e654512dde9a2cf2ad4a7a5cc36e5d44df9
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556210568 +0100

Add route for rating an article

Add route for rating an article

Setup controller for rating

Setup controller for rating

Setup controller for rating

Add route for rating an article

Setup controller for rating

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Setup controller for rating

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Setup controller for rating

Add route for rating an article

Setup controller for rating

Add route for rating an article

Setup controller for rating

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

Add route for rating an article

164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

parent 2ade73e2fee6667968edbe5feffc93db41d0fa1d
author Emmanuel Osuh <mrmusonant@gmail.com> 1555156967 +0100
committer Duziem Ugoji <tickettoduzz@gmail.com> 1556204762 +0100

Add route for rating an article

Add route for rating an article

Setup controller for rating

Add route for rating an article

Setup controller for rating

Add route for rating an article

Add route for rating an article

Setup controller for rating

Setup controller for rating

Setup controller for rating

Setup controller for rating

164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

164797150-feature(articles): merging conflicts

- merging from develop

[#164797150]

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Add route for rating an article

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Setup controller for rating

musonant committed:
Fix merge conflicts

musonant committed:
Fix merge conflicts

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

Japhethca commented on src/routes/controllers/articles.controller.js:
```diff
@@ -165,3 +166,63 @@ export async function unlikeAnArticle(req, res) {
+    }
+
+    const ratedArticle = await Article.findByPk(articleId, {
+      attributes: {
```
You can remove the `attributes` query field.

Japhethca commented on src/tests/routes/controllers/articles.test.js:
```diff
@@ -162,4 +168,43 @@ describe('Articles', () => {
+          .send({
+            rating,
+          });
+        console.log(res.body);
```
remove console.log statement.

Japhethca reviewed (changes requested):
just a few comments to address.

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

Japhethca commented on src/server.js:
```diff
@@ -28,10 +28,10 @@ const swaggerDefinition = {
-  swaggerDefinition,
-  apis: ['./**/routes/*.js'], // pass all in array
-});
+// const swaggerSpec = swaggerJSDoc({
```
remove commented code

nwamugo replied:
Okay. Done!

Japhethca commented on src/routes/middlewares/validateUser.js:
```diff
@@ -0,0 +1,43 @@
+ * @returns {Object} Error Response if validation fails
+ * @callback next when validation passes
+ */
+const validateNewUser = (req, res, next) => {
```
Am sure we have this function somewhere in the codebase, why create new one here?

Japhethca replied:
Also, I don't see where you are using this in this PR.

nwamugo replied:
It came back as an old file accidentally. It is now deleted completely.

Japhethca commented on src/routes/middlewares/validateUser.js:
```diff
@@ -0,0 +1,43 @@
+  next();
+};
+
+export const verifyUser = async (req, res, next) => {
```
Why did add this here? Lets not include what we are not using a PR.

nwamugo replied:
Old file. Deleted completely.

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

igbominadeveloper commented on src/db/seeders/20190405073145-seed-user.js:
```diff
@@ -34,8 +34,8 @@ export default {
           password: await hashPassword('smiley007'),
-          created_at: new Date().toLocaleString(),
-          updated_at: new Date().toLocaleString(),
+          created_at: new Date(),
```
I think I remember we agreed to return the .toLocaleString() here

igbominadeveloper commented on src/db/seeders/20190405073145-seed-user.js:
```diff
@@ -45,8 +45,8 @@ export default {
           password: await hashPassword('hhrtuyhgty5t678'),
-          created_at: new Date().toLocaleString(),
-          updated_at: new Date().toLocaleString(),
+          created_at: new Date(),
```
same as above

igbominadeveloper commented on src/db/seeders/20190405073145-seed-user.js:
```diff
@@ -65,8 +65,8 @@ export default {
           password: await hashPassword('nedyudombat'),
-          created_at: new Date().toLocaleString(),
-          updated_at: new Date().toLocaleString(),
+          created_at: new Date(),
```
same as above

igbominadeveloper commented on src/db/seeders/20190405073145-seed-user.js:
```diff
@@ -76,8 +76,8 @@ export default {
-          created_at: new Date().toLocaleString(),
-          updated_at: new Date().toLocaleString(),
+          created_at: new Date(),
+          updated_at: new Date(),
```
same as above

igbominadeveloper commented on src/db/seeders/20190410232221-comment-seeder.js:
```diff
@@ -9,8 +9,8 @@ export default {
           body: 'Description goes here',
-          created_at: new Date().toLocaleString(),
-          updated_at: new Date().toLocaleString(),
+          created_at: new Date(),
```
same as above

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -253,11 +257,48 @@ articleRouter.post('/:articleId/like', checkAuthorizedUser, likeAnArticle);
+articleRouter.post(
+  '/:articleId/rate',
+  checkAuthorizedUser,
+  validateRating,
```
I think we should validate the request body first before we proceed to hit the database. i.e. `validateRating` before `checkAuthorizedUser`

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -253,11 +257,48 @@ articleRouter.post('/:articleId/like', checkAuthorizedUser, likeAnArticle);
+/**
+ * @swagger
+ *
+ * /api/v1/article/:articleId/like:
```
The route documentation says `article`/:articleId/`like` but this endpoint is for unlike. It should read `articles`/:articleId/unlike` or is the like and unlike article using a single endpoint?

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -287,12 +328,85 @@ articleRouter.post('/:articleId/like', checkAuthorizedUser, likeAnArticle);
+/**
+ * @swagger
+ *
+ * /api/v1/article/:articleId/bookmark:
```
route documentation too here should be `/articles` not `article`

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -287,12 +328,85 @@ articleRouter.post('/:articleId/like', checkAuthorizedUser, likeAnArticle);
+/**
+ * @swagger
+ *
+ * /api/v1/article/:articleId/removebookmark:
```
same as above

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -287,12 +328,85 @@ articleRouter.post('/:articleId/like', checkAuthorizedUser, likeAnArticle);
+ *       401:
+ *         description: Authorization information is missing or invalid.
+ *       500:
+ *         description: Server did not process request
```
Nice one here

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -302,7 +416,7 @@ articleRouter.post(
  * @swagger
  *
- * /api/v1/article/:articleId/unlike:
+ * /api/v1/article/:articleId/like:
```
`articles` not `article`

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -416,7 +530,44 @@ articleRouter.post(
  * @swagger
  *
- * /api/v1/article/:articleId/unlike:
+ * /api/v1/article/:articleId/rate:
```
same as above

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -416,7 +530,44 @@ articleRouter.post(
+articleRouter.post(
+  '/:articleId/rate',
+  checkAuthorizedUser,
+  validateRating,
```
Kindly refer to - https://github.com/andela/zinnia-ah-backend/pull/53/files#r277587264

igbominadeveloper commented on src/routes/articles.route.js:
```diff
@@ -416,7 +530,44 @@ articleRouter.post(
+/**
+ * @swagger
+ *
+ * /api/v1/article/:articleId/like:
```
Kindly refer to - https://github.com/andela/zinnia-ah-backend/pull/53/files#r277588445

igbominadeveloper commented on src/routes/controllers/articles.controller.js:
```diff
@@ -7,6 +7,7 @@ import {
   errorResponse,
   verifyToken,
   getArticlebyId,
+  calcAverageRating,
```
I think this function can be inside the rating controller if it is not being used or re-used in some place else. Instead of having it in the global helpers. Or what do you think?

nwamugo replied:
Fair point. We can have it in a routes helper file. The function however can stand independent of the rating controller, hence its localisation.

nwamugo committed:
164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

nwamugo committed:
164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

nwamugo committed:
164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

nwamugo committed:
164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

igbominadeveloper commented on src/routes/controllers/articles.controller.js:
```diff
@@ -396,3 +397,60 @@ export async function reportArticle(req, res) {
+      });
+    }
+
+    const ratedArticle = await Article.findByPk(articleId, {
```
Great work here. Really detailed and self-explanatory.

igbominadeveloper commented on src/routes/controllers/articles.controller.js:
```diff
@@ -396,3 +397,60 @@ export async function reportArticle(req, res) {
+      res,
+      200,
+      'Your rating has been recorded',
+      jsonRatedArticle,
```
The name choice here `jsonratedarticle` sounds a bit off. Can we just call it `article`?

nwamugo replied:
I have renamed it to ratedArticleJSON

igbominadeveloper commented on src/tests/routes/controllers/articles.test.js:
```diff
@@ -309,6 +330,78 @@ describe('Articles', () => {
+        expect(res.body.data.averageRating).to.eql(4);
+      });
+
+      it('should update rating when the record already exists and but is different', async () => {
```
There is a typo in this test title

nwamugo replied:
Thank you. Corrected.

nwamugo committed:
164797150-feature(articles): merging conflicts

- merging from develop

[#164797150]

nwamugo committed:
164797150-feature(articles): merging conflicts

- merging from develop

[#164797150]

nwamugo committed:
Merge branch 'ft-user-rate-article-164797150' of github.com:andela/zinnia-ah-backend into ft-user-rate-article-164797150

nwamugo committed:
164797150-feature(articles): user can rate an article

- implement feedback as it regards correct route documentation, putting helper function into the controller, and improving the 500 test oracle

[#164797150]

nwamugo committed:
Merge branch 'develop' into ft-user-rate-article-164797150
